### PR TITLE
fix: wrong algorithms for password reset in 'settings.py'

### PR DESCRIPTION
**Pull request by rmlibre**
Password hashing occurs within the ``User.password_hash`` property setter. But, the hashing was mistakenly being done before assigning to the property, leading to an incoherent state. This PR removes the offending code, & for consistency uses the ``User`` class' password verification method.

Future work: Add tests which show password resets from 'settings.py' work as intended.

Cherry-picked from (b2cb88e) on PR branch [#411]
Merged with (ccb357d008a136eaf0d10eb761df27ab4df4a012) from 'scidsg/password-reset-fix' branch.

#### Passing workflows:

- [x] [Run Linters & Tests](https://github.com/rmlibre/hushline/actions/runs/10170023871/job/28128218992) on (https://github.com/scidsg/hushline/pull/447/commits/b10d5879333b2d6b23feee7b2db9f25f1cb73a44)
- [x] [Run Linters & Tests](https://github.com/rmlibre/hushline/actions/runs/10170330810/job/28129216156) on (https://github.com/scidsg/hushline/pull/447/commits/df0b3c08248f7813d88db71cf194cf722ffc899f)
- [x] [Run Linters & Tests](https://github.com/rmlibre/hushline/actions/runs/10172109035/job/28134183856) on (https://github.com/scidsg/hushline/pull/447/commits/0e4e5a57bcbedf3e8d4151fc2f9d9a5b50159e91)

### CODE REVIEW

File: hushline/settings.py [hunk 1]
```diff
@@ -177,9 +176,9 @@ def index() -> str | Response:  # noqa: PLR0911, PLR0912
 
             # Handle Change Password Form Submission
             if change_password_form.validate_on_submit():
-                if scrypt.verify(change_password_form.old_password.data, user.password_hash):
+                if user.check_password(change_password_form.old_password.data):
                     # Hash the new password using scrypt and update the user object
```
Comment (jeremywmoore) [inline hushline/settings.py:180]:
```suggestion
```

**Review by jeremywmoore — COMMENTED** [review]
Looks good, but let's remove the old comment since it's no longer correct or needed.

File: hushline/settings.py (hunk 1 above)
Comment (rmlibre) [reply hushline/settings.py]: Adopted by (https://github.com/scidsg/hushline/pull/447/commits/1e3f937c402750d23fddf0abe2536d90f976de6e)

**Review by jeremywmoore — APPROVED** [review]

**Review by jeremywmoore — CHANGES_REQUESTED** [review]
@rmlibre this looks like a critical PR, but I think it really needs a test. Think you could add one?

**Review by jeremywmoore — COMMENTED** [review]
One change, otherwise lgtm

File: tests/test_settings.py [hunk 2]
```diff
@@ -92,6 +94,62 @@ def test_change_username(client: FlaskClient) -> None:
...
+    new_password = f"{token_urlsafe(16)}!!!"
+    user = register_user(client, username, original_password)
+    assert user is not None, "User registration failed"
+    assert len(original_password_hash := user.password_hash) > 32
+    assert original_password_hash.startswith("$scrypt$")
+    assert original_password not in original_password_hash
+
+    # Log in the registered user
+    logged_in_user = login_user(client, username, original_password)
+    if logged_in_user is not None:
```
[the remark this answers is not in the document]
Comment (rmlibre) [reply tests/test_settings.py:110]: The ``user`` variable is checked for ``None`` on line 103. Is that the check you mean, or something else?

https://github.com/scidsg/hushline/blob/df0b3c08248f7813d88db71cf194cf722ffc899f/tests/test_settings.py#L102-L111

File: hushline/settings.py [hunk 3]
```diff
@@ -177,9 +176,8 @@ def index() -> str | Response:  # noqa: PLR0911, PLR0912
 
             # Handle Change Password Form Submission
             if change_password_form.validate_on_submit():
-                if scrypt.verify(change_password_form.old_password.data, user.password_hash):
-                    # Hash the new password using scrypt and update the user object
-                    user.password_hash = scrypt.hash(change_password_form.new_password.data)
+                if user.check_password(change_password_form.old_password.data):
```
Comment (rmlibre) [inline hushline/settings.py:179]: A new password shouldn't be allowed if it's the same as the old password.

**Review by rmlibre — COMMENTED** [review]
Another fix.

File: tests/test_settings.py (hunk 2 above)
[the remark this answers is not in the document]
Comment (jeremywmoore) [reply tests/test_settings.py]: You got it 👍

**Review by jeremywmoore — APPROVED** [review]

File: hushline/settings.py (hunk 3 above)
Comment (jeremywmoore) [reply hushline/settings.py:179]: Agreed, let's pick this up in a follow-on PR or open an issue for later.

**Pull request merged**